src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -1505,39 +1505,27 @@ public static bool CheckSchemeName(string? schemeName)
             (uint)(digit - 'a') <= 'f' - 'a' ? digit - 'a' + 10 :
             throw new ArgumentException(nameof(digit));
 
-        //
-        // GetHashCode
-        //
-        //  Overrides default function (in Object class)
-        //
-        //
         public override int GetHashCode()
         {
             if (IsNotAbsoluteUri)
             {
-                return CalculateCaseInsensitiveHashCode(OriginalString);
-            }
-
-            // Consider moving hash code storage from m_Info.MoreInfo to m_Info
-            UriInfo info = EnsureUriInfo();
-            if ((object?)info.MoreInfo == null)
-            {
-                info.MoreInfo = new MoreInfo();
+                MoreInfo moreInfo = (_info ??= new UriInfo()).MoreInfo ??= new MoreInfo();
+                if (moreInfo.Hash == 0)
+                {
+                    moreInfo.Hash = OriginalString.GetHashCode();
+                }
+                return moreInfo.Hash;
             }
-            int tempHash = info.MoreInfo.Hash;
-            if (tempHash == 0)
+            else
             {
-                string? chkString = info.MoreInfo.RemoteUrl;
-                if ((object?)chkString == null)
-                    chkString = GetParts(UriComponents.HttpRequestUrl, UriFormat.SafeUnescaped);
-                tempHash = CalculateCaseInsensitiveHashCode(chkString);
-                if (tempHash == 0)
+                MoreInfo moreInfo = EnsureUriInfo().MoreInfo ??= new MoreInfo();
```

Comment:
@stephentoub - how far down the rabbit hole do we want to go here with regard to null coalescing syntax? This is a clever application but I admit it might affect maintainability of code since it involves null coalescing _plus_ a possible assignment to a field on an object (not this) that's returned as part of a function call _plus_ a dereference of that field, all in the same line.

Reply:
I tried to not take it [too far](https://gist.github.com/MihaZupan/7b16cd33e7230f61a3ebc688a78a1a98) :)

Reply:
With #33042, it becomes `MoreInfo moreInfo = EnsureUriInfo().MoreInfo;`